VS 1036 documenting quotas (#8649)
broadinstitute/gatk: scripts/variantstore/docs/aou/workspace/CreateWriteStreamRequestIncreasedQuota.md

Case description:
In the project terra-e40eac26 we were using the BigQuery Write API to load data into BigQuery and had many failures with the message "RESOURCE_EXHAUSTED: Exceeds 'CreateWriteStream requests' quota, user_id: project0000002c6ed4d198_us (status: INSUFFICIENT_TOKENS), you can issue a raise quota request through Google Cloud Console." 
When I used the Google Cloud Console to investigate, the Quota "CreateWriteStream requests quota for us region per minute," showed 0% usage for the past seven days. We are contacting you to get this quota increased.
Note that this support ticket is just like 46139869, for which I believe the ultimate outcome was a quota increase: "Our product team have completed the quota change to 30K per 4 hours based on our discussion".
Also note that our usage for this project is very bursty, we anticipate having a high volume of CreateWriteStream requests for several days and then it should die down to a much more limited amount.

## Calculate Quota To be Requested
The default quota for this value is 50. The number that you will want for the quota is likely closer to 500 and will require some calculations to land on precisely.

> The Calculations initially got us to 500 Write Streams—for two streams
> Currently, however there are three WriteStreams that we use. One for variant information (into the vet_xxx tables) and a second for reference information (into the ref_ranges_xxx tables) and lastly a sample information stream that will put the headers into place.

Reviewer: ```suggestion
The calculations initially got us to 500 write streams for two distinct types of stream. Currently, however, there are three possible types of write streams that we use. One for variant information (into the vet_xxx tables), a second for reference information (into the ref_ranges_xxx tables) and lastly a sample information stream that will put the headers into place (if that feature is turned on).
```

## Request Additional Quota for Bulk Ingest
BigQuery Storage API CreateWriteStream quota for small regions per minute per region for region : us-central1

> (the region for our project is us-central1 so that is the quota we needed to request an update on)

Reviewer: nit
```suggestion
(the region for our project is us-central1 so that is the quota we needed to request an update on).
```

> Note that once the quota has been adjusted, it will not necessarily be visible as anything other than the default on this page.

Reviewer: nit
```suggestion
**Note:** that once the quota has been adjusted, it will not necessarily be visible as anything other than the default on this page.
```

|---------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `CreateWriteStream` requests  | 10,000 streams every hour, per project per region | You can call CreateWriteStream up to 10,000 times per hour per project per region. Consider using the default stream if you don't need exactly-once semantics. |
For more information, see [Google's quota documentation](https://cloud.google.com/bigquery/quotas)

## File a Google Support Ticket
You can point to [this ticket](https://console.cloud.google.com/support/cases/detail/v2/47548796?project=broad-dsde-methods) as an example. It is summarized below.

> Case title: In the project terra-e40eac26 we are hitting theBigQuery Write API quota "RESOURCE_EXHAUSTED"..." (see below in details)
> Category: Technical > BigQuery > API - Storage API and Streaming
> Project ID: projects/broad-dsde-methods
> Observed error message:
> com.google.api.gax.rpc.ResourceExhaustedException: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: Exceeds 'CreateWriteStream requests' quota, user_id: project0000002c6ed4d198_us (status: INSUFFICIENT_TOKENS), you can issue a raise quota request through Google Cloud Console. 

Reviewer: nit
```suggestion
- Case title: In the project terra-e40eac26 we are hitting theBigQuery Write API quota "RESOURCE_EXHAUSTED"..." (see below in details)
- Category: Technical > BigQuery > API - Storage API and Streaming
- Project ID: projects/broad-dsde-methods
- Observed error message:
com.google.api.gax.rpc.ResourceExhaustedException: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: Exceeds 'CreateWriteStream requests' quota, user_id: project0000002c6ed4d198_us (status: INSUFFICIENT_TOKENS), you can issue a raise quota request through Google Cloud Console. 
```

> Be sure to include this full error message in the request description. Before requesting an increase, please ensure your existing streams have enough utilization in terms of write traffic. For guidelines see https://cloud.google.com/bigquery/docs/write-api-best-practices#limit_the_rate_of_stream_creation. Entity: projects/aou-genomics-curation-prod/datasets/delcho_v2/tables/ref_ranges_001
> 
> Case description:
> In the project terra-e40eac26 we were using the BigQuery Write API to load data into BigQuery and had many failures with the message "RESOURCE_EXHAUSTED: Exceeds 'CreateWriteStream requests' quota, user_id: project0000002c6ed4d198_us (status: INSUFFICIENT_TOKENS), you can issue a raise quota request through Google Cloud Console." 

Reviewer: more formatting
```suggestion
- Be sure to include this full error message in the request description. Before requesting an increase, please ensure your existing streams have enough utilization in terms of write traffic. For guidelines see https://cloud.google.com/bigquery/docs/write-api-best-practices#limit_the_rate_of_stream_creation. Entity: projects/aou-genomics-curation-prod/datasets/delcho_v2/tables/ref_ranges_001
- Case description: In the project terra-e40eac26 we were using the BigQuery Write API to load data into BigQuery and had many failures with the message "RESOURCE_EXHAUSTED: Exceeds 'CreateWriteStream requests' quota, user_id: project0000002c6ed4d198_us (status: INSUFFICIENT_TOKENS), you can issue a raise quota request through Google Cloud Console." 
```

Also note that our usage for this project is very bursty, we anticipate having a high volume of CreateWriteStream requests for several days and then it should die down to a much more limited amount.

## Calculate Quota To be Requested
The default quota for this value is 50. The number that you will want for the quota is likely closer to 500 and will require some calculations to land on precisely.
This is the value that will be input for `load_data_batch` and is necessary to input for large amounts of data.

The calculations initially got us to 500 write streams for two distinct types of stream. 
Currently, however, there are three possible types of write streams that we use. One for variant information (into the vet_xxx tables), a second for reference information (into the ref_ranges_xxx tables) and lastly a sample information stream that will put the headers into place (if that feature is turned on).

Google has told us that it would be technically feasible to apply 45k per 4 hours, but we've settled on 30k per 4 hours.

For Echo:
We calculated that the `load_data_batch` should be set to 1245 without a quota increase.

> In Echo there are 414,830 non-control samples that need to be loaded during the bulk ingest process.

Reviewer 2: If you're going to document this calculation here, which I don't think you necessarily need to, you'll probably want to document an important part that you missed.  We first had to determine how many genomes are loaded by a single thread in an hour by examining the logs for one thread.  That gave us 10-15 minutes as an estimate, and we actually started our calculations from there.  That's actually how we got the theoretical CreateWriteStream usage of a single thread per hour, and then calculated how many threads we could support based upon that.

Author: huh--I just assumed that most users would have the same amount of data in a GVCF, but I guess that totally goes out the window if they are using exomes or something!
